### ISSUE TITLE
How to scope a directory in Rails starting from Rails 7.1.2?

### ISSUE THREAD

**Issue by kalsan** [rails/rails#50066]
I have a directory `app/components` that should contain classes within a module `Components`. For instance:

```ruby
# app/components/users/show.rb

class Components::Users::Show
end
```

Up and including Rails 7.1.1, this could be achieved by using:

```ruby
Rails.application.config.eager_load_paths.delete(Rails.root.join('app', 'components').to_s)
Rails.application.config.eager_load_paths.unshift(Rails.root.join('app').to_s)
```

However, in Rails 7.1.2, despite this being a patch release, this is broken and Rails expects me to have the class named `Users::Show` instead of `Components::Users::Show`.

I found that 7.1.2 also significantly alters `Rails.application.config.eager_load_paths`:

In Rails 7.1.1 (situation WITHOUT modifying `eager_load_paths`):

```
Rails.application.config.eager_load_paths
=> 
["/home/me/git/myrailsapp/app/channels",
 "/home/me/git/myrailsapp/app/components",
 "/home/me/git/myrailsapp/app/controllers",
 "/home/me/git/myrailsapp/app/controllers/concerns",
 "/home/me/git/myrailsapp/app/helpers",
 "/home/me/git/myrailsapp/app/jobs",
 "/home/me/git/myrailsapp/app/mailers",
 "/home/me/git/myrailsapp/app/models",
 "/home/me/git/myrailsapp/app/models/concerns"]
```

In Rails 7.1.2:

```
Rails.application.config.eager_load_paths
=> ["/home/me/git/myrailsapp/app", "/home/me/git/myrailsapp/lib"]
```

By intuition, I'd see this smells like a bug, because it appeared in a patch version. But perhaps it's intended and fixes the bug of it not having changed in 7.1.1. So if it is a bug, this shall be my bug report. Otherwise, I'd like to ask:

How to scope a directory in Rails starting from Rails 7.1.2, such that I can have a class  `Components::Users::Show` in `app/components/users/show.rb`?

Best,
Kalsan

**Comment by kalsan** [comment]
Likely caused by pull request #49636 to fix #49629

**Comment by fxn** [comment]
Could you please have a look at the docs for custom namespaces? https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#custom-namespaces

**Comment by fxn** [comment]
To keep things simple, I believe that could go just under the `config/initializers` folder of the engine itself.

However, your initializer should work, there has to be some detail somewhere. Could you please throw

```
Rails.autoloaders.log!
```

right after loading the defaults in `config/application.rb` and trigger the error?

**Comment by kalsan** [comment]
Hi Xavier, thanks for following up on this, much appreciated.

I think I figured out my issue. The error occurs when my initializer looks as follows:

```ruby
      unless Rails.application.config.eager_load
        Rails.application.config.to_prepare do
          Rails.autoloaders.main.push_dir("#{Rails.root}/app/components", namespace: ::Components)
          Rails.autoloaders.main.eager_load_dir("#{Rails.root}/app/components")
        end
      end
```

Correct is:

```ruby
      Rails.autoloaders.main.push_dir("#{Rails.root}/app/components", namespace: ::Components)
      unless Rails.application.config.eager_load
        Rails.application.config.to_prepare do
          Rails.autoloaders.main.eager_load_dir("#{Rails.root}/app/components")
        end
      end
```

Appearently, I mixed that up in the many tries that led to the above post.

It also works fine when the initializer is placed under the gem's `config/initializers/some_initializer.rb` directory.

So to round this up, it appears that I had used an outdated way of configuring namespaces, as the API for this has considerably improved in Rails 7.1. This deprecated way was appearently removed in 7.1.2 (wherease in 7.1.1 it worked fine), causing a patch version to break my mechanism, which should probably have been broken by 7.1.0.

In summary, I think this behavior is correct (even though suprising at first, but there won't be many others affected by it).

In my opinion, this issue can be closed. It will hopefully help the few others who will fall into this sinkhole ;-)

Thank you very much for your support and have a great day!

Kalsan

**Comment by fxn** [comment]
@kalsan yeah, I am so sorry this broke your app in a patch release.

It was a perfect storm. Rails has had a bug lurking for decades that was surfaced by 7.1 and was affecting people. It had to be fixed in a patch release, it is a bug.

The contract for the `config.*` paths was ill-defined. For example, `config.autoload_paths` is empty! How come? What can you do with these config endpoints? Unclear!

What was documented was that you could _push_ to them. So, I had to fix the bug, and did it in a way that preserved that property (pushing works). Would have loved to fix the bug and preserve mutations, but they were just opposite goals.

**Comment by kalsan** [comment]
@fxn makes sence, thanks for the context :-) yeah, it was certainly the right thing to do to clean this up. I think the new API is much more intuitive and way cleaner than the hackeries that were required for previous versions. Appreciate the effort and glad the step was finally made! I'm sure future version will offer opportunities to further refine the contract to your ideas, which definitively sound reasonable.

Cheers!
Kalsan
